src/main/java/com/browserstack/automate/ci/common/uploader/AppUploaderHelper.java

```diff
@@ -47,23 +49,36 @@ public static String uploadApp(Actionable build, PrintStream logger, String appP
 
     AppUploader appUploader = new AppUploader(appPath, credentials, customProxy, logger);
     String appId = "";
-    try {
-      PluginLogger.log(logger, "Uploading app " + appPath + " to Browserstack.");
-      appId = appUploader.uploadFile();
-      PluginLogger.log(logger,
-          appPath + " uploaded successfully to Browserstack with app_url : " + appId);
-    } catch (AppAutomateException e) {
-      PluginLogger.logError(logger, "App upload failed.");
-      PluginLogger.logError(logger, e.getMessage());
-      return null;
-    } catch (InvalidFileExtensionException e) {
-      PluginLogger.logError(logger, e.getMessage());
-      return null;
-    } catch (FileNotFoundException e) {
-      PluginLogger.logError(logger, e.getMessage());
-      return null;
+    for (int attempt = 1; attempt <= MAX_RETRY_ATTEMPTS; attempt++) {
+      try {
+        PluginLogger.log(logger, String.format("Uploading app %s to Browserstack. Attempt %d of %d", appPath, attempt, MAX_RETRY_ATTEMPTS));
+        appId = appUploader.uploadFile();
+        PluginLogger.log(logger,
+                String.format("%s uploaded successfully to Browserstack with app_url : %s", appPath, appId));
+        return appId;
+      } catch (AppAutomateException e) {
```

Comment:
Would we be retrying for all exceptions including 401 ? - invalid access key

Other concerns raised in slack DM (lets close there)

Reply:
thanks for pointing this out, we would only be retrying in case of 5XX or the status code being 0 when a request was not sent, updated the code